device/common/include/traccc/device/concepts/barrier.hpp

```diff
@@ -0,0 +1,27 @@
+/**
+ * traccc library, part of the ACTS project (R&D line)
+ *
+ * (c) 2024 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+#include "traccc/definitions/concepts.hpp"
+
+#if __cpp_concepts >= 201907L
+#include <concepts>
+#endif
+
+namespace traccc::device::concepts {
+#if __cpp_concepts >= 201907L
+template <typename T>
+concept barrier = requires(T& b) {
+    { b.blockBarrier() }
+    ->std::same_as<void>;
+    { b.blockOr(std::declval<bool>()) }
+    ->std::same_as<bool>;
+};
+#endif
```

Comment:
This is a bit subjective, but I'd rather have just a single check on the availability of concepts. I think the following should be enough:

```c++
/**
 * traccc library, part of the ACTS project (R&D line)
 *
 * (c) 2024 CERN for the benefit of the ACTS project
 *
 * Mozilla Public License Version 2.0
 */

#pragma once

#if __cpp_concepts >= 201907L

// System include(s).
#include <concepts>

/// Namespace holding all C++20 concepts used in device code
namespace traccc::device::concepts {

/// Concept that language-specific "barrier types" must fullfil
template <typename T>
concept barrier = requires(T& b) {
    { b.blockBarrier() }
    ->std::same_as<void>;
    { b.blockOr(std::declval<bool>()) }
    ->std::same_as<bool>;
};

}  // namespace traccc::device::concepts

#endif 
```

I.e. please add some documentation to the code as well. :wink:

Reply:
Yeah this has ups and downs that I was on the fence about, too. On the one hand, your solution is cleaner. On the other hand, this way it depends on the preprocessor whether the `traccc::device::concepts` namespace exists. In my experience errors about non-existent namespaces are super confusing, so I preferred to have the namespace exist regardless and then to fill it with concepts only if the feature is enabled... Can go either way though.